### PR TITLE
RL4J - (partial) RNN support

### PR DESCRIPTION
**Pull request by aboulang2002**
## What changes were proposed in this pull request?

This PR add a partial support for RNN. The support is partial because, for the time being, only whole-episode updates are supported for `NStepQLearning` and `AdvantageActorCritic`. And there is no support for RNN in the case of replay-memory based algos (DQN).
One significant difference between dense layers networks and RNNs is the presence of an internal state. In the update algorithms, when we do the feedforward to compute the expected future value and the values at each experience, special care must be taken to ensure that this internal state is correct. This is why only whole-episodes are supported with this PR (we need the whole trajectory to have the correct state); more work needs to be done (in a future PR) to save/set the neural net's internal state at each batch update (or each experience element in the case of replay-memory algos)

I added a new environment, `TMazeEnvironment`, based on https://papers.nips.cc/paper/1953-reinforcement-learning-with-long-short-term-memory.pdf, with minor changes to make the learning faster.
Use the included `TMazeExample` class to test. According to the paper above, we should consider the environment solved if 80% of the runs are successes. The sync n-step q-learning should get to that within ~2300 episodes, and ~1000 episodes in the case of A2C.
In the case of async A3C, the example doesn't always converge within 3000 episodes. I suspect the gradients are not 'sparse' enough since we're using whole-episode batches. I.e. there are too many cases of a thread overwriting the updates of another. We'll see if this is fixed when I make the changes to support non-whole-episodes updates. Also, I have on my to-do list a 'batched' neural net updater (multi threads agents, but sync updates; a bit like `AsyncGlobal` works) that would also solve the problem.

## How was this patch tested?

Manual tests and unit tests

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X] Created tests for any significant new code additions.
- [X] Relevant tests for your changes are passing.

### CODE REVIEW

**Comment by aboulang2002** [comment]
Last commit:
I realized that `EpsGreedy`, when doing a random action did not call the wrapped policy. This prevented the RNN from seeing all observations and it's internal state was incorrect.
Changing `EpsGreedy` to always call it's wrapped policy in the case of RNNs made the TMazeEnvironment converge in 1300 episodes instead of 2300 when using NStepQLearning.

**Comment by saudet** [comment]
Please fix the merge conflicts! Thanks

File: rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                        .setInputTypes(InputType.recurrent(NUM_INPUTS))
+                        .addInputs("input")
+                        .addLayer("goal", new LSTM.Builder()
+                                .nOut(40)
+                                .activation(Activation.TANH)
+                                .build(), "input")
+                        .addLayer("corridor", new DenseLayer.Builder().nOut(40).activation(Activation.RELU).build(), "input", "goal")
+                        .addLayer("corridor-1", new DenseLayer.Builder().nOut(20).activation(Activation.RELU).build(), "corridor")
+                        .addVertex("corridor-rnn", new PreprocessorVertex(new FeedForwardToRnnPreProcessor()), "corridor-1");
+    }
```
Comment (saudet) [inline rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java:165]: Does this mean users will need to build neural networks on their own? There won't be any defaults?

**Comment by saudet** [comment]
You've started mixing Windows-style line endings and UNIX-style line endings. Please decide which one you'd like to use and stick with it because GitHub isn't able to deal with them. I recommend UNIX-style line endings.

File: rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java (hunk 1 above)
Comment (aboulang2002) [reply rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java:165]: Users will indeed need to build their own neural nets. Since every environment is different, a user will need to use his expertise and intuition the same way he would if he created a network for a new deep learning task. I think DL4J has tools to help create networks? Like adding a 'convolutional block' to a network? These will be useful. And also, I have a (long term) objective to be able to use Arbiter with RL4j, to help fine-tune the network's hyperparameters but also the reinforcement learning hyperparameters (gamma, training batch size, ...)

We could however offer defaults networks for our environments 'wrappers' (GYM, ALE, Malmo, ...)

**Comment by aboulang2002** [comment]
> You've started mixing Windows-style line endings and UNIX-style line endings. Please decide which one you'd like to use and stick with it because GitHub isn't able to deal with them. I recommend UNIX-style line endings.

Sorry about that. I had to reformat my PC some time ago and reinstalled IntelliJ. The default is CR-LF and I forgot to change it back to LF.

Also, there is something strange in the master branch. Some changes were lost; for example `rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/agent/learning/algorithm/AdvantageActorCritic.java` has been deleted in a recent commit. (See https://github.com/KonduitAI/deeplearning4j/commit/0f2ccee11ed72c61401599b8c46add3a6d604ac3#diff-298c2f4683db9b1b05c458d9ed9b96b3)
What would be the best way to fix it? I propose that I change the line endings back to LF (90 files) and overwrite this bad merge in rl4j.

File: rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java (hunk 1 above)
Comment (saudet) [reply rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java:165]: Users shouldn't need to define custom networks, no. Of course, we need to be able to add new networks, but it shouldn't be required. They should also not be associated with the environments. They need to be independent, well, actually part of the configuration of the RL algorithms, not of the environments.

**Comment by saudet** [comment]
> Also, there is something strange in the master branch. Some changes were lost; for example `rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/agent/learning/algorithm/AdvantageActorCritic.java` has been deleted in a recent commit. (See [0f2ccee#diff-298c2f4683db9b1b05c458d9ed9b96b3](https://github.com/KonduitAI/deeplearning4j/commit/0f2ccee11ed72c61401599b8c46add3a6d604ac3#diff-298c2f4683db9b1b05c458d9ed9b96b3))
> What would be the best way to fix it? I propose that I change the line endings back to LF (90 files) and overwrite this bad merge in rl4j.

@agibsonccc Please fix your broken commit.

**Comment by agibsonccc** [comment]
Thanks @saudet , not sure what happened there. Can you link to a last known good state (here I'm assuming: https://github.com/KonduitAI/deeplearning4j/commit/72f5c18c830f62df2c04fbf8dc7b1353cc2d3182) and I'll revert that.

**Comment by aboulang2002** [comment]
It's not that simple unfortunately. As you know, some environments require RNNs, other convolutions, others will do just fine with dense layers, etc... And the type of each layer and how they are connected together will have an enormous impact on how well they train. For example, this is the network that solves the TMaze environment:

image

This environment require the network to remember the first observation until the end of the maze. This is why the LSTM is connected to the output.
Now, if we use instead CartPole and we remove the angular momentum from the state.  The network will also need a LSTM to approximate the missing information, but if we use the same topology as for TMaze, it will train very slowly (or not at all) because the direct link between the LSTM and the output is detrimental. If we just remove that link, like this:

image

The new network learns much faster. 

Don't get me wrong, having default networks would be very nice. But I havent't found how we should do it.
Do you have an idea?

**Comment by aboulang2002** [comment]
@agibsonccc Yes, the commit in your comment is the last know good state.

File: rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java (hunk 1 above)
Comment (aboulang2002) [reply rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java:165]: It's not that simple unfortunately. As you know, some environments require RNNs, other convolutions, others will do just fine with dense layers, etc... And the type of each layer and how they are connected together will have an enormous impact on how well they train. For example, this is the network that solves the TMaze environment:

image

This environment require the network to remember the first observation until the end of the maze. This is why the LSTM is connected to the output.
Now, if we use instead CartPole and we remove the angular momentum from the state.  The network will also need a LSTM. But this time, to approximate the missing information. If we use the same topology as for TMaze, it will train very slowly (or not at all) because the direct link between the LSTM and the output is detrimental. If we just remove that link, like this:

image

The new network learns much faster. 

Don't get me wrong, having default networks would be very nice. But I havent't found how we should do it.
Do you have an idea?

Comment (saudet) [reply rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java:165]: I'm not saying it shouldn't support custom networks, I'm saying there should be reasonable defaults for each algorithm.

**Comment by agibsonccc** [comment]
@saudet @aboulang2002 I think I got it. No more merge conflicts. Thanks guys!

File: rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java (hunk 1 above)
Comment (aboulang2002) [reply rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java:165]: I understand and agree. I think we should add several 'default' (or more realistically, a few builders that can be configured into several networks). A user will be able to test several different topologies and either pick one that works well or at least have a good starting point to design his own. But if you don't mind it to be done later, I'd like to finish the recurrent support, then add the training logging support, and add load/save support before the defaults.

Comment (saudet) [reply rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java:165]: Ok, but how are the defaults going to work? It pretty much needs to be in some configuration for each algorithm somewhere. For now, let's at least decide how that's going to look like?

Comment (aboulang2002) [reply rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java:165]: Ok, here are my first thoughts:
The first thing is to move the network to the configuration. The configuration should support being passed either directly a network instance, or an instance of `Builder<ITrainableNeuralNet>`.

If a `Builder<ITrainableNeuralNet>` is used, and since the output of a network is determined by the algorithm used, we could be nice and offer an option to automatically add the correct outputs to the network.

Instead of just offering default network, we could also offer a way to build simple networks. With this simple network builder, simple blocks are stacked such as:
- DenseLayerBlock: a block of N `DenseLayer`, with an optional max-pool at the end.
- CNNBlock: a block of N `ConvolutionLayer`, also with an optional max-pool at the end.
- RNNBlock: a block of N `LSTM` layers

Within a block, the size of individual layers can be annealed to create a funnel-like block (e.g. 20-10-5), a square block (e.g. 20-20-20) or a inverse-funnel (e.g. 5-10-20)
A block is always connected to the output of the preceding block, and optionally directly connected to the network output layer(s) (except the last block which is always connected to the network output, obviously).

Finally, we can offer a few default networks, which will internally use the above builder, such as:
- A default network for image based environments
- A default network for memory-based networks or POMDPs (a recurrent net.)
- A default network for fully observable state networks (dense layers only)

What do you think?

Comment (saudet) [reply rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java:165]: Yes, something like that should work. There should still be defaults like there are for all other configuration parameters though.
